src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultImportDocumentManager.cs

```diff
@@ -0,0 +1,253 @@
+            _templateEngineFactoryService = templateEngineFactoryService;
+            _foregroundDispatcher = foregroundDispatcher;
+            _errorReporter = errorReporter;
+
+            _importTrackerCache = new Dictionary<string, ImportTracker>(StringComparer.OrdinalIgnoreCase);
+        }
+
+        public override void OnSubscribed(VisualStudioDocumentTracker tracker)
+        {
+            if (tracker == null)
+            {
+                throw new ArgumentNullException(nameof(tracker));
+            }
+
+            _foregroundDispatcher.AssertForegroundThread();
+
+            var imports = GetImportItems(tracker);
+            foreach (var import in imports)
+            {
+                var importFilePath = import.PhysicalPath;
+                Debug.Assert(importFilePath != null);
+
+                if (!_importTrackerCache.TryGetValue(importFilePath, out var importTracker))
+                {
+                    // First time seeing this import. Start tracking it.
+                    importTracker = new ImportTracker(importFilePath);
+                    _importTrackerCache[importFilePath] = importTracker;
+
+                    StartListeningForChanges(importTracker);
+                }
+
+                importTracker.AssociatedDocuments.Add(tracker.FilePath);
+            }
+        }
+
+        public override void OnUnsubscribed(VisualStudioDocumentTracker tracker)
+        {
+            if (tracker == null)
+            {
+                throw new ArgumentNullException(nameof(tracker));
+            }
+
+            _foregroundDispatcher.AssertForegroundThread();
+
+            var imports = GetImportItems(tracker);
+            foreach (var import in imports)
+            {
+                var importFilePath = import.PhysicalPath;
+                Debug.Assert(importFilePath != null);
+
+                if (_importTrackerCache.TryGetValue(importFilePath, out var importTracker))
+                {
+                    importTracker.AssociatedDocuments.Remove(tracker.FilePath);
+
+                    if (importTracker.AssociatedDocuments.Count == 0)
+                    {
+                        // There are no open documents that care about this import. We no longer need to track it.
+                        StopListeningForChanges(importTracker);
+                        _importTrackerCache.Remove(importFilePath);
+                    }
+                }
+            }
+        }
+
+        private IEnumerable<RazorProjectItem> GetImportItems(VisualStudioDocumentTracker tracker)
+        {
+            var projectDirectory = Path.GetDirectoryName(tracker.ProjectPath);
+            var templateEngine = _templateEngineFactoryService.Create(projectDirectory, _ => { });
+            var imports = templateEngine.GetImportItems(tracker.FilePath);
+
+            return imports;
+        }
+
+        private void FireImportChanged(string importPath, ImportChangeKind kind)
+        {
```

Comment:
even though we assert it earlier, lets just assert foreground thread here prior to firing this event to make it extra clear.